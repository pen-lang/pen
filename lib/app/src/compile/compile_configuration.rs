pub struct CompileConfiguration {
    pub heap: HeapConfiguration,
    pub list_type: ListTypeConfiguration,
}

pub type HeapConfiguration = fmm_llvm::HeapConfiguration;
pub type ListTypeConfiguration = lang::hir_mir::ListTypeConfiguration;
