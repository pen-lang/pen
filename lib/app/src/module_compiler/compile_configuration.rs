pub struct CompileConfiguration {
    pub instruction: InstructionConfiguration,
    pub list_type: ListTypeConfiguration,
    pub string_type: StringTypeConfiguration,
    pub error_type: ErrorTypeConfiguration,
}

pub type InstructionConfiguration = fmm_llvm::InstructionConfiguration;
pub type ListTypeConfiguration = hir_mir::ListTypeConfiguration;
pub type StringTypeConfiguration = hir_mir::StringTypeConfiguration;
pub type ErrorTypeConfiguration = hir_mir::ErrorTypeConfiguration;
