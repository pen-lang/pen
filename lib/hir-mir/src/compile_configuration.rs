use super::{
    error_type_configuration::ErrorTypeConfiguration,
    list_type_configuration::ListTypeConfiguration, map_type_configuration::MapTypeConfiguration,
    string_type_configuration::StringTypeConfiguration,
};
#[cfg(test)]
use super::{
    error_type_configuration::ERROR_TYPE_CONFIGURATION,
    list_type_configuration::LIST_TYPE_CONFIGURATION,
    map_type_configuration::MAP_TYPE_CONFIGURATION,
    string_type_configuration::STRING_TYPE_CONFIGURATION,
};
#[cfg(test)]
use once_cell::sync::Lazy;

#[derive(Clone, Debug)]
pub struct CompileConfiguration {
    pub list_type: ListTypeConfiguration,
    pub map_type: MapTypeConfiguration,
    pub string_type: StringTypeConfiguration,
    pub error_type: ErrorTypeConfiguration,
    pub spawn_function_name: String,
    pub debug_function_name: String,
}

#[cfg(test)]
pub static COMPILE_CONFIGURATION: Lazy<CompileConfiguration> = Lazy::new(|| CompileConfiguration {
    list_type: LIST_TYPE_CONFIGURATION.clone(),
    map_type: MAP_TYPE_CONFIGURATION.clone(),
    string_type: STRING_TYPE_CONFIGURATION.clone(),
    error_type: ERROR_TYPE_CONFIGURATION.clone(),
    spawn_function_name: "spawn".into(),
    debug_function_name: "debug".into(),
});
