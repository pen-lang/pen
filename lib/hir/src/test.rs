mod ir;
mod position;

pub use self::position::position;
pub use ir::*;
