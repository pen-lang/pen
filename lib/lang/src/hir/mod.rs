pub mod analysis;
mod argument;
mod arithmetic_operation;
mod boolean;
mod boolean_operation;
mod call;
mod declaration;
mod definition;
mod else_branch;
mod equality_operation;
mod expression;
mod if_;
mod if_list;
mod if_type;
mod if_type_branch;
mod lambda;
mod let_;
mod list;
mod list_element;
mod module;
mod none;
mod not_operation;
mod number;
mod operation;
mod order_operation;
mod record_construction;
mod record_deconstruction;
mod record_element;
mod record_update;
mod string;
mod try_operation;
mod type_alias;
mod type_coercion;
mod type_definition;
mod variable;

pub use argument::*;
pub use arithmetic_operation::*;
pub use boolean::*;
pub use boolean_operation::*;
pub use call::*;
pub use declaration::*;
pub use definition::*;
pub use else_branch::*;
pub use equality_operation::*;
pub use expression::*;
pub use if_::*;
pub use if_list::*;
pub use if_type::*;
pub use if_type_branch::*;
pub use lambda::*;
pub use let_::*;
pub use list::*;
pub use list_element::*;
pub use module::*;
pub use none::*;
pub use not_operation::*;
pub use number::*;
pub use operation::*;
pub use order_operation::*;
pub use record_construction::*;
pub use record_deconstruction::*;
pub use record_element::*;
pub use record_update::*;
pub use string::*;
pub use try_operation::*;
pub use type_alias::*;
pub use type_coercion::*;
pub use type_definition::*;
pub use variable::*;
