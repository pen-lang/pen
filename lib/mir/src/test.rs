mod ir;

pub use ir::*;
