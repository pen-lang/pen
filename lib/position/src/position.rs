use crate::format;
use serde::{Deserialize, Serialize};
use std::{
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    rc::Rc,
};

#[derive(Clone, Debug, Deserialize, Serialize)]
pub struct Position(Rc<PositionInner>);

#[derive(Clone, Debug, Deserialize, Serialize)]
struct PositionInner {
    path: String,
    line_number: usize,
    column_number: usize,
    line: String,
}

impl Position {
    pub fn new(
        path: impl Into<String>,
        line_number: usize,
        column_number: usize,
        line: impl Into<String>,
    ) -> Self {
        Self(
            PositionInner {
                path: path.into(),
                line_number,
                column_number,
                line: line.into(),
            }
            .into(),
        )
    }

    pub fn path(&self) -> &str {
        &self.0.path
    }

    pub fn line_number(&self) -> usize {
        self.0.line_number
    }

    pub fn column_number(&self) -> usize {
        self.0.column_number
    }

    pub fn line(&self) -> &str {
        &self.0.line
    }
}

impl Display for Position {
    fn fmt(&self, formatter: &mut Formatter) -> fmt::Result {
        write!(formatter, "{}", format(self))
    }
}

impl Eq for Position {}

impl PartialEq for Position {
    fn eq(&self, _: &Self) -> bool {
        true
    }
}

impl Hash for Position {
    fn hash<H: Hasher>(&self, _: &mut H) {}
}

impl Ord for Position {
    fn cmp(&self, _: &Self) -> Ordering {
        Ordering::Equal
    }
}

impl PartialOrd for Position {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

#[cfg(test)]
mod tests {
    use super::Position;

    #[test]
    fn display() {
        assert_eq!(
            format!("{}", Position::new("file", 1, 1, "x")),
            "file\n1:1:\tx\n    \t^"
        );

        assert_eq!(
            format!("{}", Position::new("file", 1, 2, " x")),
            "file\n1:2:\t x\n    \t ^"
        );
    }
}
