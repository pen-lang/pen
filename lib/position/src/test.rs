use crate::Position;

pub trait PositionFake {
    fn fake() -> Self;
}

impl PositionFake for Position {
    fn fake() -> Self {
        Self::new("", 1, 1, "")
    }
}
