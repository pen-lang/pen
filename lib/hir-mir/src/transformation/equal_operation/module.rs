use super::{function, operation};
use crate::{context::CompileContext, error::CompileError, transformation::collection_type};
use hir::{
    ir::*,
    types::{self, Type},
};

pub fn transform(context: &CompileContext, module: &Module) -> Result<Module, CompileError> {
    Ok(Module::new(
        module.type_definitions().to_vec(),
        module.type_aliases().to_vec(),
        module.foreign_declarations().to_vec(),
        module.function_declarations().to_vec(),
        module
            .function_definitions()
            .iter()
            .cloned()
            .chain(
                collection_type::collect_comparable_parameter_types(context, module)?
                    .into_iter()
                    .map(|type_| compile_function_definition(context, &type_))
                    .collect::<Result<Vec<_>, _>>()?,
            )
            .collect(),
        module.position().clone(),
    ))
}

fn compile_function_definition(
    context: &CompileContext,
    type_: &Type,
) -> Result<FunctionDefinition, CompileError> {
    const LHS_NAME: &str = "$lhs";
    const RHS_NAME: &str = "$rhs";

    let position = type_.position();
    let name = function::transform_name(type_, context.types())?;

    Ok(FunctionDefinition::new(
        &name,
        &name,
        Lambda::new(
            vec![
                Argument::new(LHS_NAME, types::Any::new(position.clone())),
                Argument::new(RHS_NAME, types::Any::new(position.clone())),
            ],
            types::Boolean::new(position.clone()),
            IfType::new(
                LHS_NAME,
                Variable::new(LHS_NAME, position.clone()),
                vec![IfTypeBranch::new(
                    type_.clone(),
                    IfType::new(
                        RHS_NAME,
                        Variable::new(RHS_NAME, position.clone()),
                        vec![IfTypeBranch::new(
                            type_.clone(),
                            operation::transform(
                                context,
                                type_,
                                &Variable::new(LHS_NAME, position.clone()).into(),
                                &Variable::new(RHS_NAME, position.clone()).into(),
                                position,
                            )?,
                        )],
                        None,
                        position.clone(),
                    ),
                )],
                None,
                position.clone(),
            ),
            position.clone(),
        ),
        None,
        false,
        position.clone(),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::compile_configuration::COMPILE_CONFIGURATION;
    use hir::test::{ModuleFake, TypeAliasFake};
    use position::{test::PositionFake, Position};

    fn transform_module(module: &Module) -> Result<Module, CompileError> {
        transform(
            &CompileContext::new(module, Some(COMPILE_CONFIGURATION.clone())),
            module,
        )
    }

    #[test]
    fn transform_comparable_type() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::List::new(types::None::new(Position::fake()), Position::fake()),
            false,
            false,
        )]);

        insta::assert_debug_snapshot!(transform_module(&module));
    }

    #[test]
    fn do_not_transform_any() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::List::new(types::Any::new(Position::fake()), Position::fake()),
            false,
            false,
        )]);

        assert_eq!(transform_module(&module), Ok(module.clone()));
    }

    #[test]
    fn do_not_transform_function() {
        let module = Module::empty().set_type_aliases(vec![TypeAlias::fake(
            "a",
            types::List::new(
                types::Function::new(vec![], types::None::new(Position::fake()), Position::fake()),
                Position::fake(),
            ),
            false,
            false,
        )]);

        assert_eq!(transform_module(&module), Ok(module.clone()));
    }
}
