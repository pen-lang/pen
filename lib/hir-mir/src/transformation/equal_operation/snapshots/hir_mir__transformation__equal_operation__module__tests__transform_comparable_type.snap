---
source: lib/hir-mir/src/transformation/equal_operation/module.rs
expression: transform_module(&module)
---
Ok(
    Module {
        type_definitions: [],
        type_aliases: [
            TypeAlias {
                name: "a",
                original_name: "",
                type_: List(
                    List {
                        element: None(
                            None {
                                position: Position {
                                    path: "",
                                    line_number: 1,
                                    column_number: 1,
                                    line: "",
                                },
                            },
                        ),
                        position: Position {
                            path: "",
                            line_number: 1,
                            column_number: 1,
                            line: "",
                        },
                    },
                ),
                public: false,
                external: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        foreign_declarations: [],
        function_declarations: [],
        function_definitions: [
            FunctionDefinition {
                name: "hir:equal:d409c28886113c9b",
                original_name: "hir:equal:d409c28886113c9b",
                lambda: Lambda {
                    arguments: [
                        Argument {
                            name: "$lhs",
                            type_: Any(
                                Any {
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                        },
                        Argument {
                            name: "$rhs",
                            type_: Any(
                                Any {
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                        },
                    ],
                    result_type: Boolean(
                        Boolean {
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    body: IfType(
                        IfType {
                            name: "$lhs",
                            argument: Variable(
                                Variable {
                                    name: "$lhs",
                                    position: Position {
                                        path: "",
                                        line_number: 1,
                                        column_number: 1,
                                        line: "",
                                    },
                                },
                            ),
                            branches: [
                                IfTypeBranch {
                                    type_: None(
                                        None {
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                    expression: IfType(
                                        IfType {
                                            name: "$rhs",
                                            argument: Variable(
                                                Variable {
                                                    name: "$rhs",
                                                    position: Position {
                                                        path: "",
                                                        line_number: 1,
                                                        column_number: 1,
                                                        line: "",
                                                    },
                                                },
                                            ),
                                            branches: [
                                                IfTypeBranch {
                                                    type_: None(
                                                        None {
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                    expression: Boolean(
                                                        Boolean {
                                                            value: true,
                                                            position: Position {
                                                                path: "",
                                                                line_number: 1,
                                                                column_number: 1,
                                                                line: "",
                                                            },
                                                        },
                                                    ),
                                                },
                                            ],
                                            else_: None,
                                            position: Position {
                                                path: "",
                                                line_number: 1,
                                                column_number: 1,
                                                line: "",
                                            },
                                        },
                                    ),
                                },
                            ],
                            else_: None,
                            position: Position {
                                path: "",
                                line_number: 1,
                                column_number: 1,
                                line: "",
                            },
                        },
                    ),
                    position: Position {
                        path: "",
                        line_number: 1,
                        column_number: 1,
                        line: "",
                    },
                },
                foreign_definition_configuration: None,
                public: false,
                position: Position {
                    path: "",
                    line_number: 1,
                    column_number: 1,
                    line: "",
                },
            },
        ],
        position: Position {
            path: "",
            line_number: 1,
            column_number: 1,
            line: "",
        },
    },
)
